pytorch/translation/openmpi-wmt14-gnmt-all-reduce/main.py

```diff
@@ -0,0 +1,414 @@
+
+    return fp_optimizer, model
+
+
+def build_criterion(padding_idx, smoothing):
+    if smoothing == 0.0:
+        criterion = nn.CrossEntropyLoss(ignore_index=padding_idx, size_average=False)
+    else:
+        criterion = LabelSmoothing(padding_idx, smoothing)
+
+    return criterion
+
+
+def train_loop(
+        run_id,
+        dataset_dir,
+        ckpt_run_dir,
+        output_dir,
+        validation_only=False,
+        use_cuda=False,
+        light_target=False,
+):
+    """Train loop"""
+    train_epochs = 6
+
+    # Dataset attributes
+    train_min_len, train_max_len = 0, 50
+    val_min_len, val_max_len = 0, 150
+    math_mode = "fp16"  # One of `fp16`, `fp32`
+    batch_first = False
+    include_lengths = True
+    lang = {"src": "en", "trg": "de"}
+
+    # Model attributes
+    hidden_size = 1024
+    num_layers = 4
+    dropout = 0.2
+    share_embedding = True
+    smoothing = 0.1
+
+    # Training
+    train_batch_size = 128
+    train_global_batch_size = 1024 if dist.get_world_size() <= 8 else 2048
+    train_iter_size = set_iter_size(train_global_batch_size, train_batch_size)
+    val_batch_size = 64
+    validate_every = 2000
+
+    assert (validate_every % train_iter_size) == 0
+    # Translator
+    beam_size = 5
+    len_norm_factor = 0.6
+    cov_penalty_factor = 0.1
+    len_norm_const = 5.0
+    max_seq_len = 150
+
+    # Optimizer
+    lr = 2.00e-3
+    grad_clip = 5.0
+
+    # Loss
+    loss_scaling = {"init_scale": 8192, "upscale_interval": 128}
+
+    # Scheduler
+    scheduler_config = {
+        "warmup_steps": 200,
+        "remain_steps": 0.666,
+        "decay_interval": None,
+        "decay_steps": 4,
+        "decay_factor": 0.5,
+    }
+
+    rank = dist.get_rank()
+    world_size = dist.get_world_size()
+
+    # Build train/val datsets
+    train_set = WMT14Dataset(
+        dataset_dir,
+        batch_first=batch_first,
+        include_lengths=include_lengths,
+        math_precision=math_mode,
+        lang=lang,
+        train=True,
+        download=True,
+        lazy=True,
+        min_len=train_min_len,
+        max_len=train_max_len,
+    )
+
+    val_set = WMT14Dataset(
+        dataset_dir,
+        batch_first=batch_first,
+        include_lengths=include_lengths,
+        math_precision=math_mode,
+        lang=lang,
+        validation=True,
+        download=False,
+        min_len=val_min_len,
+        max_len=val_max_len,
+    )
+    tokenizer = train_set.fields["trg"]
+
+    # Build model
+    model = GNMT(
+        vocab_size=train_set.vocab_size,
+        hidden_size=hidden_size,
+        num_layers=num_layers,
+        dropout=dropout,
+        batch_first=batch_first,
+        share_embedding=share_embedding,
+    )
+
+    # Build loss function
+    loss_function = build_criterion(config.PAD, smoothing)
+
+    # Bilingual Evaluation Understudy Score
+    metrics = [BLEUScore()]
+
+    # Partition data
+    train_set = partition_dataset_by_rank(train_set, rank, world_size)
```

Comment:
If you look e.g. here https://github.com/mlbench/mlbench-benchmarks/blob/develop/pytorch/imagerecognition/openmpi-cifar10-resnet20-all-reduce/main.py#L84
we also partition the validation set to the different workers, so each worker has to only do validation on 1/#workers part of the validation set. 
Then `validation_round` aggregates all the validations here https://github.com/mlbench/mlbench-core/blob/2f52d54de9608549087ae7ae6760544d3c554d4d/mlbench_core/controlflow/pytorch/controlflow.py#L197 
This speeds up validation a bit (and after allreduce in the train loop, all workers are in the same state). Minor optimization but still cuts down on total time to a result a bit.